compiler/src/main/kotlin/io/spine/protodata/backend/DependencyEvents.kt

```diff
@@ -0,0 +1,194 @@
+import io.spine.protodata.Service
+import io.spine.protodata.enumType
+import io.spine.protodata.event.dependencyDiscovered
+import io.spine.protodata.messageType
+import io.spine.protodata.name
+import io.spine.protodata.oneofGroup
+import io.spine.protodata.path
+import io.spine.protodata.service
+import io.spine.protodata.typeUrl
+
+/**
+ * Creates a `DependencyDiscovered` event from the given file descriptor.
+ *
+ * The event reflects all the definitions from the file.
+ */
+internal fun discoverDependencies(fileDescriptor: FileDescriptor) =
+    dependencyDiscovered {
+        val id = fileDescriptor.path()
+        file = id
+        content = fileDescriptor.toPbSourceFile()
+    }
+
+private fun FileDescriptor.toPbSourceFile(): ProtobufSourceFile {
+    val path = path()
+    val result = ProtobufSourceFile.newBuilder()
+        .setFilePath(path)
+        .setFile(toFileWithOptions())
+    val doc = Documentation.fromFile(this)
+    with(DefinitionsBuilder(path, doc)) {
+        result.putAllType(
+            messageTypes()
+                .map { it.name.typeUrl() to it }
+                .toMap()
+        )
+        result.putAllEnumType(
+            enumTypes()
+                .map { it.name.typeUrl() to it }
+                .toMap()
+        )
+        result.putAllService(
+            services()
+                .map { it.name.typeUrl() to it }
+                .toMap()
+        )
+    }
+    return result.build()
+}
+
+/**
+ * A builder for the Protobuf definitions of a single `.proto` file.
+ *
+ * @property path the relative file path to the Protobuf file
+ * @property documentation all the documentation and comments present in the file
+ */
+private class DefinitionsBuilder(
+    private val path: FilePath,
+    private val documentation: Documentation,
+) {
+
+    /**
+     * Builds the message type definitions.
+     *
+     * @return all the message types declared in the file, including nested types.
+     */
+    fun FileDescriptor.messageTypes(): Sequence<MessageType> {
+        var messages = messageTypes.asSequence()
+        for (msg in messageTypes) {
+            messages += walkMessage(msg) { it.nestedTypes }
+        }
+        return messages.map { it.asMessage() }
+    }
+
+    /**
+     * Builds the enum type definitions.
+     *
+     * @return all the enums declared in the file, including nested enums.
+     */
+    fun FileDescriptor.enumTypes(): Sequence<EnumType> {
+        var enums = enumTypes.asSequence()
+        for (msg in messageTypes) {
+            enums += walkMessage(msg) { it.enumTypes }
+        }
+        return enums.map { it.asEnum() }
+    }
+
+    /**
+     * Builds the service definitions.
+     *
+     * @return all the services declared in the file, including the nested ones.
+     */
+    fun FileDescriptor.services(): Sequence<Service> =
+        services.asSequence().map { it.asService() }
+
+    private fun Descriptor.asMessage() = messageType {
```

Comment:
This method deserves documentation. It looks complicated. Please also consider refactoring.